### PR TITLE
Minor improvements for knob framework

### PR DESCRIPTION
**Pull request by angela28chen**
* Replace `GetFlagHelpText()` with storing help parameters separately and assembling in `KnobManager::GetUsageMsg()`
* Move `KnobCheckbox` member functions to `knob.cpp`
* When listing `KnobDropdown` choices in usage message, only add quotation marks if there are spaces
* Made KnobManager a friend class of Knob and removed some of the getters/setters, preferring to have protected/private members where possible
* Renamed `SetFlagDesc` -> `SetFlagDescription`

### CODE REVIEW

File: include/ppx/knob.h [hunk 1]
```diff
@@ -54,16 +54,15 @@ class Knob
     const std::string& GetFlagName() const { return mFlagName; }
     const std::string& GetFlagHelp() const { return mFlagDesc; }
     size_t             GetIndent() const { return mIndent; }
+    const std::string& GetHelpParams() const { return mHelpParams; }
```
Comment (apazylbe) [inline include/ppx/knob.h:57]: `GetHelpParams` does not sound very descriptive. It would help to have a more descriptive name or at least a comment. Perhaps we could also group the three flag functions together and give them related names. `GetFlagHelp` above is also a bit confusing, and potentially misleading as well (since there was `GetFlagHelpText` below.
Perhaps `GetFlagName`, `GetFlagParameters`/`GetFlagValues` and `GetFlagDescription`?

File: include/ppx/knob.h [hunk 2]
```diff
@@ -54,16 +54,15 @@ class Knob
...
 
     void SetDisplayName(const std::string& displayName) { mDisplayName = displayName; }
     void SetFlagDesc(const std::string& flagDesc) { mFlagDesc = flagDesc; }
     void SetIndent(size_t indent) { mIndent = indent; }
+    void SetHelpParams(const std::string& helpParams) { mHelpParams = helpParams; }
```
Comment (apazylbe) [inline include/ppx/knob.h:62]: Looking at just changes here and not how this will be used later, how do you feel about overriding `GetHelpParams` to build the string instead of having setter/getter. Perhaps later usage of SetHelpParams changes things though

File: include/ppx/knob.h [hunk 3]
```diff
@@ -54,16 +54,15 @@ class Knob
...
 
     virtual void Draw() = 0;
 
-    // Returns the flag description used in usage message
-    virtual std::string GetFlagHelpText() const = 0;
```
Comment (apazylbe) [inline include/ppx/knob.h:65]: Are `GetFlagName`, `GetFlagHelp` and `GetHelpParams` only used by the KnobManager to build help text?
I  think you moved this function out of Knob because you want to align the description?
You could consider making the KnobManager a friend and keeping some of this interface private, if it doesn't need to be public.

File: src/test/knob_test.cpp [hunk 4]
```diff
@@ -371,15 +365,15 @@ TEST_F(KnobManagerTestFixture, KnobManager_GetUsageMsg)
...
     std::shared_ptr<KnobDropdown<std::string>> strKnobPtr1(km.CreateKnob<KnobDropdown<std::string>>("flag_name4", 1, choices1.cbegin(), choices1.cend()));
 
     std::string usageMsg = R"(
-Application-specific flags
+Application-Specific Flags:
 --flag_name1 <true|false>
 --flag_name2 <true|false>
 --flag_name3 <0~10>
---flag_name4 <"c1"|"c2"|"c3">
+--flag_name4 <c1|c2|"c3 and more">
```
Comment (apazylbe) [inline src/test/knob_test.cpp:376]: Curious, why this change?

File: include/ppx/knob.h [hunk 5]
```diff
@@ -243,6 +202,21 @@ class KnobDropdown final
         : Knob(flagName), mIndex(defaultIndex), mDefaultIndex(defaultIndex), mChoices(choicesBegin, choicesEnd)
     {
         PPX_ASSERT_MSG(defaultIndex < mChoices.size(), "defaultIndex is out of range");
+        std::string choiceStr = "";
+        for (const auto& choice : mChoices) {
+            bool hasSpace = choice.find_first_of("\t ") != std::string::npos;
+            if (hasSpace) {
+                choiceStr += R"(")" + choice + R"("|)";
```
Comment (Keenuts) [inline include/ppx/knob.h:209]: nit: why use the more complex `R"( )"` notation vs `"\""` ?

**Review by Keenuts — APPROVED** [review]
Ok with this, @apazylbe has covered remaining issues, delegating approval 😊

**Review by apazylbe — CHANGES_REQUESTED** [review]

**Review by apazylbe — APPROVED** [review]
